src/LeaseManager.cc

```diff
@@ -0,0 +1,307 @@
+/* Copyright (c) 2014-2015 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <string>
+
+#include "LeaseManager.h"
+
+#include "ExternalStorage.h"
+#include "LeaseCommon.h"
+#include "ShortMacros.h"
+
+namespace RAMCloud {
+
+/// Defines the number of leases the preallocator should try to keep ahead of
+/// the lastIssuedLeaseId.
+const uint64_t PREALLOCATION_LIMIT = 1000;
+
+/// Defines the prefix for objects stored in external storage by this module.
+const std::string STORAGE_PREFIX = "leaseManager";
+
+LeaseManager::LeaseManager(Context* context)
+    : mutex()
+    , context(context)
+    , clock(context)
+    , lastIssuedLeaseId(0)
+    , maxAllocatedLeaseId(0)
+    , leaseMap()
+    , expirationOrder()
+    , preallocator(context, this)
+    , cleaner(context, this)
+{}
+
+/**
+ * Return the lease info for the provided leaseId.  Used by masters to
+ * implicitly check if a lease is still valid.
+ *
+ * \param leaseId
+ *      Id of the lease whose liveness you wish to check.
+ * \return
+ *      If the lease exists, return lease with its current lease term.  If not,
+ *      the returned the lease id will be 0 (an invalid lease id) signaling that
+ *      the requested lease has already expired.
+ */
+WireFormat::ClientLease
+LeaseManager::getLeaseInfo(uint64_t leaseId)
+{
+    Lock lock(mutex);
+    WireFormat::ClientLease clientLease;
+
+    // Populate the lease information if it is found.
+    LeaseMap::iterator leaseEntry = leaseMap.find(leaseId);
+    if (leaseEntry != leaseMap.end()) {
+        clientLease.leaseId = leaseId;
+        clientLease.leaseTerm = leaseEntry->second;
```

Comment:
Looks like you are leaving  one field of the lease uninitialized. However, I don't see any documentation for that possibility in WireFormat::ClientLease.